src/libertem/udf/base.py

```diff
@@ -1777,54 +1935,71 @@ def run_for_dataset_sync(
+                    for res in executor.run_tasks(
                         tasks,
                         params_handle,
                         cancel_id,
-                    )
-                    for part_results, task in result_iter:
+                    ):
                         if progress:
                             t.update(1)
-                        for results, udf in zip(part_results, self._udfs):
-                            udf.set_views_for_partition(task.partition)
-                            udf.merge(
-                                dest=udf.results.get_proxy(),
-                                src=results.get_proxy()
-                            )
-                            udf.clear_views()
-                        v = damage.get_view_for_partition(task.partition)
-                        v[:] = True
-                        yield UDFResults(
-                            buffers=tuple(
-                                udf._do_get_results()
-                                for udf in self._udfs
-                            ),
-                            damage=damage
-                        )
-                else:
-                    # yield at least one result (which should be empty):
-                    for udf in self._udfs:
-                        udf.clear_views()
-                    yield UDFResults(
-                        buffers=tuple(
-                            udf._do_get_results()
-                            for udf in self._udfs
-                        ),
-                        damage=damage
-                    )
+                        yield res
         finally:
             if progress:
                 t.close()
 
+    def run_for_dataset_sync(
+        self,
+        dataset: DataSet,
+        executor: JobExecutor,
+        roi: Optional[np.ndarray] = None,
+        progress: bool = False,
+        corrections: Optional[CorrectionSet] = None,
+        backends: Optional[BackendSpec] = None,
+        dry: bool = False,
+        iterate: bool = True
+    ) -> Generator["UDFResults", None, None]:
+        executor = executor.ensure_sync()
+        result_iter = self.results_for_dataset_sync(
+            dataset=dataset,
+            executor=executor,
+            roi=roi,
+            progress=progress,
+            corrections=corrections,
+            backends=backends,
+            dry=dry,
+        )
+        damage = BufferWrapper(kind='nav', dtype=bool)
+        damage.set_shape_ds(dataset.shape, roi)
+        damage.allocate()
+        any_result = False
+        for part_results, task in result_iter:
+            any_result = True
+            res = self._apply_part_result(
+                udfs=self._udfs,
+                damage=damage,
+                part_results=part_results,
+                task=task
+            )
+            # Explicit indexing for compatibility with Dask.delayed
+            self._udfs = res[0]
+            damage = res[1]
```

Comment:
Are these three lines really still necessary? If so, please remind me why, and we probably should add a test case that demonstrates this

Reply:
Yes, not necessary anymore thanks to @matbryan52 's work! :-)